### ISSUE TITLE
CMakeClose fail to close the console window

### ISSUE THREAD

**Issue by BlurryLight** [cdelledonne/vim-cmake#25]
**Describe the bug**
CMakeClose fail to close the console window

**To Reproduce**
1. `CMakeBuild` or `CMakeGenerate`,  which write some chars into the console window buffer. Or just call `CMakeOpen` and input some random chars.
2. call `CMakeClose` . *Error37: No write since last change (add ! to override)*

**Expected behavior**
- `CMakeClose` should close the window even if there is change in the buffer. 
- Another option is to have a command `CMakeClose!` , which has the similar behaviour like `:q!` and keep the current `CMakeClose` for those who want to save the CMake log from the console window.

**Screenshots**
![图片](https://user-images.githubusercontent.com/14940338/103147807-a3df0a00-4793-11eb-973a-8d2cef08e105.png)

**Other info**
- OS: Manjaro x64
- Vim/Neovim version: vim 8.2
- Vim-CMake version (5th line in `:help cmake`): 0.4.0

**Additional context**

**Comment by BlurryLight** [comment]
https://github.com/cdelledonne/vim-cmake/blob/09c043adde3be5dba499fef68ea27d7209d8db07/autoload/cmake/console.vim#L185

Modifying this line into `execute l:cmake_window . 'close!' `  works as what I expected.

**Comment by cdelledonne** [comment]
Hi, thanks a lot for posting this.

At the moment I cannot look into it very carefully, I should be able to do it in the next days though. Out of curiosity, with your fix, what happens when trying to close the CMake window while the CMake window is the last window?

**Comment by BlurryLight** [comment]
It will be the standard vim behaviour: "E444: Cannot close the last window".  Users have to use `:q!` to quit the vim.

**Comment by matu3ba** [comment]
Even on closing, I get an infinity loop taking up a full CPU.

**Comment by akorchemniy** [comment]
Running :CMakeBuild also switches the active window into insert mode. Kinda of annoy. Have to remember to leave insert mode, then switch to the CMake console window, leave terminal mode, then cq.

**Comment by cdelledonne** [comment]
Apologies for the delay on this, at the moment I am quite busy with some other stuff and cannot work on this plugin too much.

I can't reproduce this issue with my Vim instance. It might be some Vim terminal-related configuration which is conflicting with this plugin. I mostly use Neovim, thus I'm not an expert on Vim's terminal options. I will look into that as soon as I find some time.

By the way, have you tried reproducing with a minimal `.vimrc`? Do you set any explicit terminal-related configuration options in your `.vimrc` file?

**Comment by cdelledonne** [comment]
> Running :CMakeBuild also switches the active window into insert mode. Kinda of annoy. Have to remember to leave insert mode, then switch to the CMake console window, leave terminal mode, then cq.

@akorchemniy I suspect this is also related to some Vim terminal-related configuration. Could you please open a separate issue though? Thanks.

**Comment by bitwisemanager** [comment]
Hey guys, is it possible automatically to close vim when CMake window is the last window?

**Comment by zeertzjq** [comment]
I tried this plugin in Neovim 5.0 Nightly and `:CMakeBuild` also switches the active window into insert mode.

**Comment by cdelledonne** [comment]
Hmm interesting, I'll try to reproduce that asap

**Comment by dilawar** [comment]
I am facing the same issue as @zeertzjq . Running neovim nightly.

**Comment by cdelledonne** [comment]
@zeertzjq @dilawar I can't reproduce your issue with Neovim nightly. I'm not sure, but it could be some configuration option in your `vimrc` file or some other plugin's option causing the issue. Anyway, please open a separate issue with a clear description of how to reproduce your problem, so that I can try to help you further. Also please try to reproduce the issue with a minimal `vimrc`

**Comment by cdelledonne** [comment]
@vnepogodin by "same error" you mean what's in the title of the issue? Or is it that running a command switches the active window into insert mode?

**Comment by dilawar** [comment]
> @vnepogodin by "same error" you mean what's in the title of the issue? Or is it that running a command switches the active window into insert mode?

 running a command switches the active window into insert mode.

**Comment by vnepogodin** [comment]
@cdelledonne same as in the title of the issue

**Comment by zeertzjq** [comment]
Oh, it seems to be caused by a plugin indeed. Sorry for bothering you.

**Comment by dilawar** [comment]
> Oh, it seems to be caused by a plugin indeed. Sorry for bothering you

which plugin? may be I have the same plaugin causing issue.

**Comment by vnepogodin** [comment]
> @zeertzjq @dilawar I can't reproduce your issue with Neovim nightly. I'm not sure, but it could be some configuration option in your `vimrc` file or some other plugin's option causing the issue. Anyway, please open a separate issue with a clear description of how to reproduce your problem, so that I can try to help you further. Also please try to reproduce the issue with a minimal `vimrc`

I've tested with minimal setup:
``` vim
syntax enable
set nu

set exrc

" Plugin Manager
call plug#begin('~/.vim/plugged')

Plug 'cdelledonne/vim-cmake'
  let g:cmake_generate_options = ['-DCMAKE_EXPORT_COMPILE_COMMANDS=1']
  let g:cmake_build_dir_location = 'build'
  let g:cmake_build_options = ['--parallel 10']

call plug#end()
```
Not working in Vim v8.2, working in Neovim v0.4.4

**Comment by cdelledonne** [comment]
Thank you for reporting this. I created a Manjaro VM (the OP's system) and I could indeed reproduce this issue. Will do some digging and get back to you  guys soon. Thank you for your patience :)

**Comment by cdelledonne** [comment]
Hey @BlurryLight @vnepogodin, I have just pushed a fix that I think should address this issue. Could you please try to pull the latest `master` and let me know if the problem went away or not?

Thank you very much in advance.

**Comment by cdelledonne** [comment]
> Hey guys, is it possible automatically to close vim when CMake window is the last window?

@bitwisemanager the `filetype` of the CMake window buffer is `vimcmake`, i.e. `&filetype == 'vimcmake'` returns `1`. You could take inspiration from this answer to implement what you need: https://stackoverflow.com/questions/7069927/in-vimscript-how-to-test-if-a-window-is-the-last-window

**Comment by vnepogodin** [comment]
> Hey @BlurryLight @vnepogodin, I have just pushed a fix that I think should address this issue. Could you please try to pull the latest `master` and let me know if the problem went away or not?
> 
> Thank you very much in advance.
I've just tested and issue is fixed. Thank you!

**Comment by BlurryLight** [comment]
@bitwisemanager 

I hope this snippet could help you.
```
aug vim-cmake-group
    au WinEnter * if winnr('$') == 1 && &filetype == "vimcmake"|q!|endif
aug END
```

@cdelledonne 
I've tested the latest fix on vim8.2 from Manjaro x64. It works as expected. Thanks for that. 

There are some discussions on another problem about switching the active window into insert mode. I believe that behaviour has little relation about the problem described in the title. It is better to open a new issue to fully discuss it and to find the possible solution. Since the problen in the issue title has been fixed, I am going to close the issue.
